#Taller 1
1.	Usando el fichero bolsa_2021.csv. Realiza los modelos CAPM de Cie, Enagás y Santander y responde a las siguientes preguntas.

Recuerda que para ello primero tendrás que generar el tipo de interés diario, las rentabilidades de cada acción (e Ibex) y las primeras de riesgo de cada acción e Ibex

a)	¿Cuáles de las acciones son defensivas y cuáles ofensivas?

b)	¿Cumple la constante lo que dice el modelo teórico CAPM?

c)	¿Cuánto del riesgo de la acción explicamos con el riesgo del mercado en cada caso?



In [1]:
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import pandas as pd
import yfinance as yf

In [2]:
df = pd.read_csv('C:/Users/Lenovo/Programacion/Python/BBDD/bolsa_2021.csv', sep=",", index_col=0)
df

Cie         Ibez     Enagas  Santander  interes
Date                                                             
19/10/2020  17.049999  6860.200195  20.100000   1.584892    0.120
20/10/2020  17.020000  6927.299805  20.100000   1.628783    0.145
21/10/2020  17.200001  6811.500000  20.100000   1.599075    0.165
22/10/2020  17.370001  6796.600098  20.100000   1.617092    0.183
23/10/2020  17.530001  6893.399902  20.100000   1.671333    0.170
...               ...          ...        ...        ...      ...
12/10/2021  21.860001  8935.900391  19.629999   3.383500    0.441
13/10/2021  22.040001  8881.400391  19.385000   3.276000    0.443
14/10/2021  22.020000  8925.000000        NaN   3.281000    0.458
15/10/2021  22.440001  8997.000000        NaN   3.345500    0.464
18/10/2021  22.360001  8970.200195        NaN   3.362000    0.492

[257 rows x 5 columns]

In [3]:
stock = df[["Cie", "Ibez", "Enagas", "Santander"]]
retornos_stock = stock.pct_change().dropna()
retornos_stock

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_2304\1901437492.py:2: FutureWarning: The default fill_method='pad' in DataFrame.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  retornos_stock = stock.pct_change().dropna()


Cie      Ibez    Enagas  Santander
Date                                               
20/10/2020 -0.001759  0.009781  0.000000   0.027693
21/10/2020  0.010576 -0.016716  0.000000  -0.018239
22/10/2020  0.009884 -0.002187  0.000000   0.011267
23/10/2020  0.009211  0.014242  0.000000   0.033542
26/10/2020 -0.003423 -0.013999  0.000000  -0.012156
...              ...       ...       ...        ...
12/10/2021 -0.009066  0.004147  0.006409   0.010302
13/10/2021  0.008234 -0.006099 -0.012481  -0.031772
14/10/2021 -0.000907  0.004909  0.000000   0.001526
15/10/2021  0.019074  0.008067  0.000000   0.019659
18/10/2021 -0.003565 -0.002979  0.000000   0.004932

[256 rows x 4 columns]

In [4]:
interes = df[["interes"]]["20/10/2020":"18/10/2021"]
interes.dropna()
y =interes/36000
y

interes
Date                
20/10/2020  0.000004
21/10/2020  0.000005
22/10/2020  0.000005
23/10/2020  0.000005
26/10/2020  0.000004
...              ...
12/10/2021  0.000012
13/10/2021  0.000012
14/10/2021  0.000013
15/10/2021  0.000013
18/10/2021  0.000014

[256 rows x 1 columns]

In [13]:
x= pd.concat([retornos_stock, y], axis=1)
x


Cie      Ibez    Enagas  Santander   interes
Date                                                         
20/10/2020 -0.001759  0.009781  0.000000   0.027693  0.000004
21/10/2020  0.010576 -0.016716  0.000000  -0.018239  0.000005
22/10/2020  0.009884 -0.002187  0.000000   0.011267  0.000005
23/10/2020  0.009211  0.014242  0.000000   0.033542  0.000005
26/10/2020 -0.003423 -0.013999  0.000000  -0.012156  0.000004
...              ...       ...       ...        ...       ...
12/10/2021 -0.009066  0.004147  0.006409   0.010302  0.000012
13/10/2021  0.008234 -0.006099 -0.012481  -0.031772  0.000012
14/10/2021 -0.000907  0.004909  0.000000   0.001526  0.000013
15/10/2021  0.019074  0.008067  0.000000   0.019659  0.000013
18/10/2021 -0.003565 -0.002979  0.000000   0.004932  0.000014

[256 rows x 5 columns]

In [14]:

x["pCie"] = x["Cie"] - x["interes"]
x["pIbez"] =  x["Ibez"] -x["interes"] 
x["pEnegas"] =  x["Enagas"] - x["interes"]
x["pSant"] = x["Santander"] - x["interes"]


In [15]:
x

Cie      Ibez    Enagas  Santander   interes      pCie  \
Date                                                                      
20/10/2020 -0.001759  0.009781  0.000000   0.027693  0.000004 -0.001764   
21/10/2020  0.010576 -0.016716  0.000000  -0.018239  0.000005  0.010571   
22/10/2020  0.009884 -0.002187  0.000000   0.011267  0.000005  0.009879   
23/10/2020  0.009211  0.014242  0.000000   0.033542  0.000005  0.009207   
26/10/2020 -0.003423 -0.013999  0.000000  -0.012156  0.000004 -0.003427   
...              ...       ...       ...        ...       ...       ...   
12/10/2021 -0.009066  0.004147  0.006409   0.010302  0.000012 -0.009078   
13/10/2021  0.008234 -0.006099 -0.012481  -0.031772  0.000012  0.008222   
14/10/2021 -0.000907  0.004909  0.000000   0.001526  0.000013 -0.000920   
15/10/2021  0.019074  0.008067  0.000000   0.019659  0.000013  0.019061   
18/10/2021 -0.003565 -0.002979  0.000000   0.004932  0.000014 -0.003579   

               pIbez   pEnegas     pSant  
Date                                      
20/10/2020  0.009777 -0.000004  0.027689  
21/10/2020 -0.016721 -0.000005 -0.018244  
22/10/2020 -0.002193 -0.000005  0.011262  
23/10/2020  0.014238 -0.000005  0.033538  
26/10/2020 -0.014003 -0.000004 -0.012160  
...              ...       ...       ...  
12/10/2021  0.004134  0.006396  0.010289  
13/10/2021 -0.006111 -0.012493 -0.031784  
14/10/2021  0.004896 -0.000013  0.001514  
15/10/2021  0.008054 -0.000013  0.019646  
18/10/2021 -0.002992 -0.000014  0.004918  

[256 rows x 9 columns]

In [16]:
est = smf.ols(formula='pCie ~ pIbez', data=x) # smf.ols modelo de regresion lineal simple (ols)
res = est.fit()
print(res.summary())
est = smf.ols(formula='pEnegas ~ pIbez', data=x)
res=est.fit()
print(res.summary())
est = smf.ols(formula='pSant ~ pIbez', data=x)
res=est.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:                   pCie   R-squared:                       0.259
Model:                            OLS   Adj. R-squared:                  0.256
Method:                 Least Squares   F-statistic:                     88.97
Date:                Fri, 01 Nov 2024   Prob (F-statistic):           2.66e-18
Time:                        14:10:16   Log-Likelihood:                 697.94
No. Observations:                 256   AIC:                            -1392.
Df Residuals:                     254   BIC:                            -1385.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0003      0.001      0.343      0.7

a)	¿Cuáles de las acciones son defensivas y cuáles ofensivas?

Debemos de revisar el dato entre (pIbez - Coef):

#### Si el valor es menor que < 1, nos indica que la Accion es defensiva
#### Si el valor es mayo a > 1 nos indica que la Accion es Ofensaiva

- Pcie = 0.7903, la Accion es defensiva por que es menos a 1.

- PEnegas = 0.0783, La Accion es defensia
    
- PSant = 1,7029, La Accion es ofensiva

b)	¿Cumple la constante lo que dice el modelo teórico CAPM?

- Pcie - No rechazamos la hipotesis nula por que en el intercept para P-valor 0.732 debe ser mayor a 0.05 si cumple el el modelo

- PEnegas - No rechazamos la hipotesis nula por que en el intercept para P-valor 0.908 debe ser mayor a 0.05

- PSant - No rechazamos la hipotesis nula por que en el intercept para P-valor 0.908 debe ser mayor a 0.05

c)	¿Cuánto del riesgo de la acción explicamos con el riesgo del mercado en cada caso?


    
- Pcie - El 25.9%  del comportamiento se puede explicar con el indice. El restando es el riesgo propio ( ideocincratico) de la acción.

- Penegas - 0.04%

- Psant - 66.4%



-------------

2.	Usando el fichero viv.csv elabora un modelo de tasación del logaritmo del precio de las viviendas tasadas. 

a)	Añade primero como explicativas el logaritmo de la superficie, de la edad y de la variable estudios de los habitantes de la zona. Analiza los resultados

b)	Añade una variable para las características de la vivienda: planta, exterior, calefacción, ascensor y estado de la vivienda. Analiza los resultados

c)	Añade una variable ficticia para el año 2000 y 2001. ¿Cuánto crecieron los precios en esos dos años respecto a 1999?


In [10]:
imb = pd.read_csv('C:/Users/Lenovo/Programacion/Python/BBDD/viv.csv', index_col=0)
imb

distrito  superf  reforma  estado  edad  ascensor  extint  lprecio  \
any                                                                        
1999         1      35        2       4    50         0       1       16   
1999         1      48        4       3    60         0       1       16   
1999         1      38        3       3    90         0       1       16   
1999         1      46        4       1    90         0       1       16   
1999         1      55        3       3    35         0       1       16   
...        ...     ...      ...     ...   ...       ...     ...      ...   
2001        10     101        3       4    35         1       1       17   
2001        10      87        7       3    28         1       1       17   
2001        10      75        4       3    32         1       1       17   
2001        10      68        3       3    28         1       1       17   
2001        10      87        3       4    26         1       1       17   

      lsuperf  ledad  ati  planta  calef  estudios  lestudios  d00  d01  
any                                                                      
1999        4      4    0       0      0         8          2    0    0  
1999        4      4    0       2      0         8          2    0    0  
1999        4      5    0       2      0         8          2    0    0  
1999        4      5    0       3      0         8          2    0    0  
1999        4      4    0       3      0         8          2    0    0  
...       ...    ...  ...     ...    ...       ...        ...  ...  ...  
2001        5      4    0       7      1         8          2    0    1  
2001        4      3    0       8      1         8          2    0    1  
2001        4      3    0       2      1         8          2    0    1  
2001        4      3    0       5      1         8          2    0    1  
2001        4      3    0       5      1         8          2    0    1  

[5867 rows x 17 columns]

In [11]:
est = smf.ols(formula='lprecio ~ lsuperf+ledad+lestudios', data=imb)
res = est.fit()
print(res.summary())
est1= smf.ols(formula='lprecio ~ lsuperf+ledad+lestudios+planta+extint+calef+ascensor+estado', data=imb)
res=est1.fit()
print(res.summary())
est2= smf.ols(formula='lprecio ~ lsuperf+ledad+lestudios+planta+extint+calef+ascensor+estado+d00+d01', data=imb)
res=est2.fit()
print(res.summary())
res=est2.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:                lprecio   R-squared:                       0.361
Model:                            OLS   Adj. R-squared:                  0.361
Method:                 Least Squares   F-statistic:                     1104.
Date:                Fri, 01 Nov 2024   Prob (F-statistic):               0.00
Time:                        13:52:58   Log-Likelihood:                -3781.0
No. Observations:                5867   AIC:                             7570.
Df Residuals:                    5863   BIC:                             7597.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.8658      0.071    196.384      0.0